best-practices/sdk-principles.md

```diff
@@ -0,0 +1,226 @@
+Matt O wrote a document for comment [2023 - Product terminology and taxonomy (2023 - Apr-May)](https://docs.google.com/document/d/1IpCJL7ME0JavWOjPL0VbsAVOy7nlO-A05H4Lf3vAbqA/edit).
+In this document, various options are presented in terms of how our SDKs should evolve in terms of the core SDKs, their composability, and our new product SDKs.
+The [v2 proposal](https://docs.google.com/document/d/1IpCJL7ME0JavWOjPL0VbsAVOy7nlO-A05H4Lf3vAbqA/edit#heading=h.jx5io6mmxzz), which is the current recommendation, is presented below.
+
+This proposal ensures that we:
+
+- create broader-scoped product SDKs that expose proposed functionality as collections or attributes so as not to limit the scope of the root object i.e. spaces has space.locks, space.members, space.cursors thus ensuring it’s extensible and we could add space.foo. 
+
+- standardise how we create and release objects; 
+
+- we consistently refer to the core SDKs as clients
+
+## Core and product simplified code examples
+### Core SDK (exposes Channels & Notifications) 
+
+Our current pub/sub SDKs should be referred to as our core SDKs. It is perfectly valid for a developer to only user a core SDK given it provides a rich feature set of pub/sub+ capabilities.
+
+```typescript
+import Ably from '@ably/core';
+const client = new Ably.Realtime(clientOptions);
+```
+
+### Composable Core SDK
+
+For languages that support composability (currently only the core), we should should have just one way of adding support for features to a base client (stripped back) core, or the standard core.
+Currently we have [plugins for features like Deltas](https://ably.com/docs/channels/options/deltas?lang=javascript), and have a different approach for our current [composable and tree-shakeable Javascript SDK](https://github.com/ably/ably-js/issues/1210).
+We should standardise on this and take one approach.
+
+Example with a lightweight base client adding subscription, presence and delta support:
+
+```typescript
+import RealtimeBase from '@ably/core/realtimebase';
+import RealtimeSubscriptionMod from '@ably/core/mod/realtimesubscription';
+import RealtimePresenceMod from '@ably/core/mod/web/realtimepresence';
+import VCDiffMod from '@ably/vcdiff-decoder';
+const client = new RealtimeBase({
+  key: 'xVLyHw.0RRPmg:JBNRyepNu6YBGGi2eiPuIwzn-D2wxLnXD2CicinBeHw',
+  modules: [VCDiffMod, RealtimeSubscriptionMod, RealtimePresenceMod]
+});
+```
+
+Example with the Core client adding delta support:
+
+```typescript
+import RealtimeBase from '@ably/core';
+import VCDiffMod from '@ably/vcdiff-decoder';
+const client = new RealtimeBase({
+  key: 'xVLyHw.0RRPmg:JBNRyepNu6YBGGi2eiPuIwzn-D2wxLnXD2CicinBeHw',
+  modules: [VCDiffMod]
+});
+```
+
+### Data Sync (current proposed product name is Live Sync)
+
+```typescript
+import { Realtime } from '@ably/core';
+import DataSyncClient from '@ably/datasync';
+const rtclient = new Realtime(ABLY_API_KEY);
```

Comment:
minor thing, but should preserve empty new lines just like in the confluence doc for better readability, see how it looks [there](https://ably.atlassian.net/wiki/spaces/ENG/pages/2692874253/Client+software+SDK+principles+and+best+practice+RFC#:~:text=Data%20Sync%20(current%20proposed%20product%20name%20is%20Live%20Sync)).

should update this and other code samples that had new lines in the original doc

Reply:
done